keps/sig-api-machinery/NNNN-built-in-default/README.md

```diff
@@ -0,0 +1,302 @@
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, built-in API types are written as Go structures annotated
+with markers to add additional semantics. This method has inspired
+controller-tools who has re-used and improved on that technique for
+building CRDs. CRDs have acquired the `// +default` tag to declaratively
+define defaults and that mechanisms still doesn't exist for built-in.
+This proposal tries to solve that inconsistency.
+
+## Motivation
+
+Having a new `// +default` marker for built-in types has several
+important benefits:
+
+- It's more convenient for types authors.
+- The types and their behaviors are easier to understand since they are
+described together.
+- They are less error-prone, and prevent people from using anti-patterns
+like "cross-field dependency defaults" (default depends on another
+field), non-deterministic or complicated defaults. These patterns
+might be useful in limited circumstances, but it should be easier to do
+the right thing in most cases.
+- Removes the inconsistency with CRDs that already support this
+mechanism.
+- Solves bug where associative keys need to either be required or
+  defaulted.
+
+### Goals
+
+The goal is to add a new `// +default` tag to our current built-in Go
+IDL. That tag will be transformed into the OpenAPI `default` tag and
```

Comment:
What are the formatting rules for the +default tag value? I'd expect it to match what OpenAPI defines, roughly, or what we have in CRDs today, but might be worth referencing the primary documentation for that here and commenting about any special considerations for putting the value into a godoc tag.

Reply:
Oh, also, How malformed default tags are detected and reported back to the developer is an interesting detail.

Reply:
Yes, will solve the comment: we should probably take this exactly as the openapi takes it.
And yes, we can fail when we generate the openapi for the types.

Reply:
The default tag in OpenAPI is not what you think it is. It's not "our default". We might want `x-kubernetes-default` instead.

Also note that we will never publish the default. But of course it's fine to get it into SSA via OpenAPi.

OpenAPI uses just json for the default. kubebuilder has some other, home-grown syntax.

Reply:
I talked to Solly, we changed the format to yaml so that it's easier to format (and still allows structures through its json form). kubebuilder would probably be able to change the format since it's a different marker name anyway.

> The default tag in OpenAPI is not what you think it is. It's not "our default". We might want `x-kubernetes-default` instead.

Can you elaborate on how it's different?

This seems to be what defines a default: https://swagger.io/specification/v2/:
> Declares the value of the item that the server will use if none is provided.

That seems to match with what I think we're doing.

Reply:
> Can you elaborate on how it's different?

Trying to remember the old discussion. It might be that it was about read-only, not about default. When rereading defaults both in JSON Schema and OpenAPI I agree with you. It sounds like our default.

Reply:
There is one maybe ugly detail which makes it different to the default of our native types:

A default in a struct (= object in OpenAPI slang) for native types applies even if the whole struct is omitted (as long as the struct is not a pointer).

A default in a struct in the OpenAPI sense (and CRDs for that matter) only applies when the struct / object is specified.

This leads to this kind of constructions to workaround the latter:

```
type: object
properties:
  foo:
    type: object
    properties:
      bar:
        type: string
        default: "abc"
    default: {}
```

In other words, the implicit default of objects is `undefined`.

At least we defined the default semantics of defaults for CRDs in a way, that the outer default `{}` is applied first, and then the value for `bar` is set. I.e. the effective default is `{"foo": {"bar": "abc"}}` here.

Reply:
Problem with ^^: due to Golang unmarshal semantics, a default of a PodSepc in native types and a default of a PodSpec in a CRD will differ.

I would feel better if we operate on JSON in either case instead of generation of defaulters for native types with different semantics.

Reply:
Thanks, that's a very good comment. I've added a section that addresses this. I'll try to provide more example.

Reply:
> At least we defined the default semantics of defaults for CRDs in a way, that the outer default `{}` is applied first, and then the value for bar is set. I.e. the effective default is `{"foo": {"bar": "abc"}}` here.

That saves our lives. I think that imposing a `{}` default on non-pointer structs solves that problem.

Reply:
> I would feel better if we operate on JSON in either case instead of generation of defaulters for native types with different semantics.

I think if we follow the steps that I've specified now, operating on json or built-in types should be equivalent.